doop-api/aggregate_test.go

```diff
@@ -0,0 +1,111 @@
+	expected := mustParseJSON[doop.AggregatedReport](t, "fixtures/output-cluster1-only.json")
+	actual := AggregateReports(inputSet, BuildFilterSet(url.Values{}))
+	assert.DeepEqual(t, "AggregateReports", actual, expected)
+
+	//test a filter that does not change anything because it exactly matches what is in the report
+	filterStr := "cluster_identity.number=one&template_kind=GkFirstTemplate&constraint_name=firstconstraint&object_identity.type=production"
+	actual = AggregateReports(inputSet, BuildFilterSet(query(filterStr)))
+	assert.DeepEqual(t, "AggregateReports", actual, expected)
+
+	//test a filter that removes all clusters
+	filterStr = "cluster_identity.number=two"
+	actual = AggregateReports(inputSet, BuildFilterSet(query(filterStr)))
+	assert.DeepEqual(t, "AggregateReports", actual, doop.AggregatedReport{
+		ClusterIdentities: map[string]map[string]string{},
+		Templates:         nil,
+	})
+
+	//test several filters that remove all violations because they mismatch on each other possible level
+	//(removing violations also removes all effectively empty objects above it)
+	negativeFilters := []string{
+		"template_kind=GkSecondTemplate",
+		"constraint_name=secondconstraint",
+		"object_identity.type=qa",
+	}
+	for _, filterStr := range negativeFilters {
+		actual = AggregateReports(inputSet, BuildFilterSet(query(filterStr)))
+		assert.DeepEqual(t, "AggregateReports", actual, doop.AggregatedReport{
+			ClusterIdentities: map[string]map[string]string{
+				"cluster1": {"number": "one"},
+			},
+			Templates: nil,
+		})
+	}
+}
+
+func TestAggregateOfTwoClusters(t *testing.T) {
+	//Each of these cluster reports has exactly one violation.
+	inputSet := map[string]doop.Report{
+		"cluster1": mustParseJSON[doop.Report](t, "fixtures/input-cluster1.json"),
+		//this one can merge with cluster 1 on same violation group
+		"cluster2": mustParseJSON[doop.Report](t, "fixtures/input-cluster2.json"),
+		//this one can merge with cluster 1 on different violation group, but same constraint
+		"cluster3": mustParseJSON[doop.Report](t, "fixtures/input-cluster3.json"),
+		//this one can merge with cluster 1 on different constraint, but same template
+		"cluster4": mustParseJSON[doop.Report](t, "fixtures/input-cluster4.json"),
+	}
+
+	//test merging of structures on all levels of the report
+	expected := mustParseJSON[doop.AggregatedReport](t, "fixtures/output-both-clusters.json")
+	actual := AggregateReports(inputSet, BuildFilterSet(url.Values{}))
+	assert.DeepEqual(t, "AggregateReports", actual, expected)
+}
+
+func query(input string) url.Values {
+	result, err := url.ParseQuery(input)
+	if err != nil {
+		panic(err.Error())
+	}
+	return result
+}
+
+func mustParseJSON[T any](t *testing.T, path string) (result T) {
+	t.Helper()
+	buf, err := os.ReadFile(path)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
+	err = json.Unmarshal(buf, &result)
+	if err != nil {
+		t.Fatal(err.Error())
+	}
```

Comment:
We can't create a generic for this iferr variant?

Reply:
I sat down a while ago and tried to come up with a design for "must.Return and must.Succeed, but for unit tests". All my attempts failed because of various language constraints. I can give more details if necessary, but for now I'd like to leave it at "it's not that easy".